File: priv/simple_options_example.ex
defmodule SimpleOptionsExample do
  # Elixir allows you to put some validation into a module attributes.
  @definitions [
    name: [
      :required,
      documentation: "Node name. Must be an atom.",
      validation: &is_atom/1
    ]
  ]

  def merge_default(opts) do
    :simple_options.merge(opts, @definitions)
  end

  @doc """
  More complex option validations, especially ones that refer to
  functions defined in the same module, need to be defined inline.
  """
  def merge(opts) do
    defintions = [
      name: [
        required: true,
        documentation: "Name of the node - either an atom or a binary",
        validation: &validate_name(&1),
        error_message: "must be an atom or a binary."
      ],
      pool_size: [
        required: false,
        default: 10,
        documentation: "Pool size as an integer. Must be bigger than 0",
        validation: fn x ->
          case {is_integer(x), x > 0} do
            {true, true} -> true
            {false, _} -> :expected_integer
            {true, false} -> :expected_larger_than_zero
          end
        end
      ]
    ]

    :simple_options.merge(opts, defintions)
  end

  def validate_name(name) do
    is_atom(name) || is_binary(name)
  end
end

File: priv/simple_options_example.erl
-module(simple_options_example).

%%%-------------------------------------------------------------------
%%% @doc
%%% You can put your definitions into a macro and use to validate
%%% before you start your application.
%%% @end
%%%-------------------------------------------------------------------
-define(OPTIONS_DEFINITION,
        [{name,
          [{required, false},
           {default, local},
           {validation,
            fun (X) ->
                    is_atom(X) orelse is_binary(X)
            end},
           {documentation, <<"Name of the node - either an atom or a binary.">>},
           {error_message, <<"must be an atom or a binary">>}]},
         {pool_size,
          [required,
           {validation,
            fun (X) ->
                    case {is_integer(X), X > 0} of
                      {true, true} ->
                          true;
                      {false, _} ->
                          expected_integer;
                      {true, false} ->
                          expected_larger_than_zero
                    end
            end},
           {documentation, <<"Pool size as an integer. Must be bigger than 0">>}]}]).

-export([start/2, start_inline/2, stop/1]).

start(_Type, _Args) ->
    UserOpts = application:get_env(my_application),
    _Options = simple_options:merge(UserOpts, ?OPTIONS_DEFINITION),
    ok.

%%%-------------------------------------------------------------------
%%% @doc
%%% Or you can put them inline.
%%% @end
%%%-------------------------------------------------------------------
start_inline(_Type, _Args) ->
    Definitions = [{name,
                    [{required, false},
                     {default, local},
                     {validation,
                      fun (X) ->
                              is_atom(X) orelse is_binary(X)
                      end},
                     {documentation, <<"Name of the node - either an atom or a binary.">>},
                     {error_message, <<"must be an atom or a binary">>}]}],
    UserOpts = application:get_env(my_application),
    _Options = simple_options:merge(UserOpts, Definitions),
    ok.

stop(_State) ->
    ok.
